sql/schemachanger: zone config decomposition

fqazi opened the pull request:
1. These changes decompose the zone config and subzones, sufficiently to allow us to support updates and changes to them.
2. Add test coverage for basic decomposition of the zone config.
3. Adopt the shared zone config reading infrastructure in existing code paths.

fqazi committed:
sql/schemachanger: adopt zone config reader

Previously, the zone config reader code was only used
by the declarative schema changer. But, we can take
advantage of logic in existing zone config logic. To address
this, this patch will adopt the zone config reader in existing
logic.

Release note: None

fqazi committed:
sql/schemachanger: add tests for zone config decomposition

Previously, there was no coverage for validating the decomposition
of zone configs, inside the declarative schema changer. To address
this, this patch adds support for testing this decomposition logic
with dependency injection.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83879)
<!-- Reviewable:end -->

postamar reviewed (commented):
I know this is still in draft mode but this is close. Your test failures are caused by a missing `dev gen bazel` and changed expected round-trip counts, which aren't surprising.

> Reviewed 21 of 24 files at r1, 6 of 6 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-:-N6IerJD0WvewhGY4HPy:b-4uwfva)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/regionutils/zone_config_reader.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IfhIK10rOFuHulwqs:-N6IfhIK10rOFuHulwqt:bd89714) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/regionutils/zone_config_reader.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> package regionutils
> ```

Consider renaming this package to `zoneconfig` or something like that and move it under `pkg/sql/catalog`, assuming something similar doesn't exist already.
___
*[`pkg/sql/regionutils/zone_config_reader.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6If8Oz6nzfABXbAMgZ:-N6If8Oz6nzfABXbAMg_:b-bc4tj2) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/regionutils/zone_config_reader.go#L52)):*
> ```Go
> 		return nil, err
> 	}
> 	return &zone, nil
> ```

Why can't we query the `zones` table instead with the internal executor? If there's a good reason, it's worth a comment.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IgKFJ7fqPPD4a3ewx:-N6IgKFK7Et3ucQ17ptm:b-vdwzti) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L74)):*
> ```protobuf
>   RowLevelTTL row_level_ttl = 29 [(gogoproto.customname) = "RowLevelTTL", (gogoproto.moretags) = "parent:\"Table\""];
>   TableZoneConfig table_zone_config = 121 [(gogoproto.moretags) = "parent:\"Table, View, Sequence\""];
>   TableSubZoneConfig table_subzone_config = 122  [(gogoproto.moretags) = "parent:\"Table, Index\""];
> ```

I don't know how much work is involved or whether it would be out of scope but assuming they're not too different, consider adding support for the whole zone config hierarchy, as I believe that databases have them, and perhaps also schemas, I don't know.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 498 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6Ih303CDfEuG0UH8Re:-N6Ih303CDfEuG0UH8Rf:bs6be8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L498)):*
> ```protobuf
> }
> 
> message TableSubZoneConfig {
> ```

Should this be split into two? `TableIndexZoneConfig` keyed by `(table_id, index_id)` and `TableIndexPartitionZoneConfig` keyed by `(table_id, index_id, partition_name)`? I don't really know how zone configs work, please correct me if that's not how they work.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IgYKcAcROSZjtuI2A:-N6IgYKcAcROSZjtuI2B:b-no2w0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L501)):*
> ```protobuf
>   uint32 table_id =1 [(gogoproto.customname) = "TableID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
>   uint32 index_id = 2 [(gogoproto.customname) = "IndexID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.IndexID"];
>   uint32 zone_config_id = 3 [(gogoproto.customname) = "ZoneConfigID"];
> ```

I would leave out this field for the time being, this feels out of scope for the time being. Same applies for `TableZoneConfig`.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 503 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6Igjl55g-Wghqj80dz:-N6Igjl55g-Wghqj80e-:b-wqofkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L503)):*
> ```protobuf
>   uint32 zone_config_id = 3 [(gogoproto.customname) = "ZoneConfigID"];
>   string partition_name = 4;
> }
> ```

Is leaving the actual zone config proto out of the element deliberate? If so please add a TODO explaining why. If not, I feel it would make the `scdecomp` tests more interesting. Same applies for `TableZoneConfig`.
___
*[`pkg/sql/schemachanger/screl/attr.go` line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IhXuXAg0oJNSIDtMu:-N6IhXuXAg0oJNSIDtMv:b3rb8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/screl/attr.go#L117)):*
> ```Go
> 		rel.EntityAttr(DescID, "TypeID"),
> 	),
> 	rel.EntityMapping(t((*scpb.EnumTypeValue)(nil)),
> ```

Doesn't this change belong to a different PR?
___
*[`pkg/ccl/schemachangerccl/testdata/decomp/multiregion` line 679 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IhtGMEmqPCq22EJ1k:-N6IhtGMEmqPCq22EJ1l:babpypt) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/ccl/schemachangerccl/testdata/decomp/multiregion#L679)):*
> ```
>     partitionName: us-east3
>     tableId: 108
>     zoneConfigId: 1
> ```

Can you also add non-ccl zone config test cases in the non-ccl suite? Like adjusting the GC TTL and whatnot.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @postamar)

___
*[`pkg/sql/regionutils/zone_config_reader.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6If8Oz6nzfABXbAMgZ:-N6InVPK8hUOvXOEXZJv:b5o35q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/regionutils/zone_config_reader.go#L52)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why can't we query the `zones` table instead with the internal executor? If there's a good reason, it's worth a comment.
</blockquote></details>

Good point, I'm going to move this over to the internal executor.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IgKFJ7fqPPD4a3ewx:-N6Ind0o22elUyAjuUND:b-kienu) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L74)):*
> I don't know how much work is involved or whether it would be out of scope but assuming they're not too different, consider adding support for the whole zone config hierarchy, as I believe that databases have them, and perhaps also schemas, I don't know.
Right, part of the reason this is in the draft is because I wanted to get down into decomposing it further. I have an idea, for doing that but want to make sure my assumptions are sane first.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 498 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6Ih303CDfEuG0UH8Re:-N6InrA-8YBVpeFmyQ6o:b-2965sl) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L498)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Should this be split into two? `TableIndexZoneConfig` keyed by `(table_id, index_id)` and `TableIndexPartitionZoneConfig` keyed by `(table_id, index_id, partition_name)`? I don't really know how zone configs work, please correct me if that's not how they work.
</blockquote></details>

Subzones have an index, table and partition associated with the zone config. This data is then used to compute the spans for each subzone, which is why they gather it I think.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IgYKcAcROSZjtuI2A:-N6IonYsCE5_-7Alq4HC:b-3cwgfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L501)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I would leave out this field for the time being, this feels out of scope for the time being. Same applies for `TableZoneConfig`.
</blockquote></details>

Yes, I'm gonna drop it out for now. We don't need it for ADD COLUMN support or zone config alters.
___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 503 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6Igjl55g-Wghqj80dz:-N6IovJCAEHAForBQGOU:b-5y249o) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L503)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is leaving the actual zone config proto out of the element deliberate? If so please add a TODO explaining why. If not, I feel it would make the `scdecomp` tests more interesting. Same applies for `TableZoneConfig`.
</blockquote></details>

Yes, it's deliberate, my goal was to further decompose the internals of the proto since embedding the whole thing will make updates impossible.  i.e. The builder should not be in the business of constructing zone configs *ever*. This stuff should be left to execution because we can have transaction retires during jobs, etc..
___
*[`pkg/sql/schemachanger/screl/attr.go` line 117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IhXuXAg0oJNSIDtMu:-N6IpNSm2ViW4IYd5Oyu:b-y7o20u) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/screl/attr.go#L117)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Doesn't this change belong to a different PR?
</blockquote></details>

Accidentally pulled this in will clean it up.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @postamar)

___
*[`pkg/sql/schemachanger/scpb/elements.proto` line 74 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IgKFJ7fqPPD4a3ewx:-N6IpUxu1XR_UYggnPWr:b-9od4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/schemachanger/scpb/elements.proto#L74)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

> I don't know how much work is involved or whether it would be out of scope but assuming they're not too different, consider adding support for the whole zone config hierarchy, as I believe that databases have them, and perhaps also schemas, I don't know.
Right, part of the reason this is in the draft is because I wanted to get down into decomposing it further. I have an idea, for doing that but want to make sure my assumptions are sane first.
</blockquote></details>

Right, part of the reason this is in the draft is that I wanted to decompose it further. I have an idea, for doing that but want to make sure my assumptions are sane first.


<!-- Sent from Reviewable.io -->

fqazi committed:
pkg/schemachanger: further decompose table zone config elements

Previously, our zone config decomposition only handled,
elements for subzones at the table level. This patch will
further include zone config options at each level.

Release note: None

fqazi committed:
sql/schemachanger: decompose database level zone config

Previously, we only decomposed the table level configuration,
which was not sufficient. This patch extends the decomposition
to the database level.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/ccl/schemachangerccl/testdata/decomp/multiregion` line 679 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IhtGMEmqPCq22EJ1k:-N6OnHah0uywcjy_Bog5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/ccl/schemachangerccl/testdata/decomp/multiregion#L679)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Can you also add non-ccl zone config test cases in the non-ccl suite? Like adjusting the GC TTL and whatnot.
</blockquote></details>

Done.
___
*[`pkg/sql/regionutils/zone_config_reader.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6IfhIK10rOFuHulwqs:-N6OExjh0Zwy7d-YRKZk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/032c364ed4c86e53b24813fa83aaec5acc6f3ec5/pkg/sql/regionutils/zone_config_reader.go#L11)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider renaming this package to `zoneconfig` or something like that and move it under `pkg/sql/catalog`, assuming something similar doesn't exist already.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/zoneconfig/zone_config_options.go:
```diff
@@ -0,0 +1,189 @@
+// assignments in ALTER CONFIGURE ZONE to assignments to the member
+// fields of zonepb.ZoneConfig.
+var SupportedZoneConfigOptions = map[tree.Name]struct {
+	RequiredType *types.T
+	Setter       func(*zonepb.ZoneConfig, tree.Datum)
+	Getter       func(*zonepb.ZoneConfig) (tree.Datum, error)
+	CheckAllowed func(context.Context, EnterpriseCheckCallback, tree.Datum) error // optional
```
nit: how would you feel about making this type exported and have these fields as private fields and then have methods which call them? The `CheckAllowed` could then no-op if the optional field is not set.

For testing you can expose a constructor. Then, lastly, unexport the map and instead have a public function to do the map lookup

ajwerner commented on pkg/sql/zoneconfig/zone_config_reader.go:
```diff
@@ -0,0 +1,53 @@
+// Copyright 2022 The Cockroach Authors.
```
Nothing in this commit uses this file, consider splitting it out.

ajwerner commented on pkg/bench/rttanalysis/testdata/benchmark_expectations:
```diff
@@ -32,21 +32,21 @@ exp,benchmark
-17,DropDatabase/drop_database_1_table
-18,DropDatabase/drop_database_2_tables
-19,DropDatabase/drop_database_3_tables
+18,DropDatabase/drop_database_1_table
+20,DropDatabase/drop_database_2_tables
+22,DropDatabase/drop_database_3_tables
```
I'm okay with this for now, but this is going to hurt in the future. I think ideally we'll find a way to pre-fetch all of the descriptors and zones in a batch. Not for this PR

ajwerner commented on pkg/sql/zoneconfig/zone_config_reader.go:
```diff
@@ -13,40 +13,47 @@ package zoneconfig
-type ZoneConfigReader struct {
-	txn   *kv.Txn
-	codec keys.SQLCodec
+// Reader implements the scbuild.ZoneConfigReader interface
```
yeah, just remove this file in the first commit.

ajwerner commented on pkg/sql/schemachanger/scexec/BUILD.bazel:
```diff
@@ -74,6 +74,7 @@ go_test(
         "//pkg/sql/catalog/tabledesc",
         "//pkg/sql/parser",
         "//pkg/sql/schemachanger/scbuild",
+        "//pkg/sql/schemachanger/scdecomp",  # keep
```
why is this needed?

ajwerner reviewed (commented):
This is not too far off, but the extra round trips is sad. Can you file an issue before we merge this. I think we'll want to do something here before shipping this release.

fqazi committed:
sql/zoneconfig: refactor zone config option for use in schemachanger

Previously, the zone config options were embedded inside the sql
package making it hard for the declarative schema changer to
access these. To help support updates to zone configuration we
are going to split these off into there own package. This patch,
will start by refactoring this into the zone config package.

Release note: None

fqazi committed:
sql/schemachanger: partially decompose zone config elements

Previously, our element model did not decompose zone config
and multi-region related elements for the declarative schema
changer. These changes add support for partially decomposing these
elements, so that we can issue proper subzone related updates
when creating indexes during add column.

Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83879)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[`pkg/sql/zoneconfig/zone_config_options.go` line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6LJTaVlhGQfsMhZ46P-r4-35:-N6QY9YG7FiufrJ0dh9B:b-huxtts) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a82d9ce04ca6fa937673b8dec8355ed847d400a/pkg/sql/zoneconfig/zone_config_options.go#L35)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: how would you feel about making this type exported and have these fields as private fields and then have methods which call them? The `CheckAllowed` could then no-op if the optional field is not set.

For testing you can expose a constructor. Then, lastly, unexport the map and instead have a public function to do the map lookup
</blockquote></details>

Good point, let me tweak that part of the code
___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 37 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6PUiaXouJ4G41Uq94x-r10-37:-N6QY3Lj3GTZgyDCG9C4:b-ggpyja) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6592b671d31ab9ca4af2e622ef81bee76c37721/pkg/bench/rttanalysis/testdata/benchmark_expectations#L37)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm okay with this for now, but this is going to hurt in the future. I think ideally we'll find a way to pre-fetch all of the descriptors and zones in a batch. Not for this PR
</blockquote></details>

Let me get a issue open for it
___
*[`pkg/sql/schemachanger/scexec/BUILD.bazel` line 77 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6GfbIbnn04uJTIDIo6-r14-77:-N6QXu4T5ysDdfA_89s8:b-628li3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e02dbd67241fab2ceb0c5288e6555ca36ead5a5c/pkg/sql/schemachanger/scexec/BUILD.bazel#L77)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why is this needed?
</blockquote></details>

Left over from earlier stuff I was prototyping. Removed.

Done.
___
*[`pkg/sql/zoneconfig/zone_config_reader.go` line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6LJTaVlhGQfsMhZ46Q-r4-1:-N6QY1zCAAwYKUkKDtrI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a82d9ce04ca6fa937673b8dec8355ed847d400a/pkg/sql/zoneconfig/zone_config_reader.go#L1)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Nothing in this commit uses this file, consider splitting it out. 
</blockquote></details>

Done.
___
*[`pkg/sql/zoneconfig/zone_config_reader.go` line 25 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/83879#-N6LJTaVlhGQfsMhZ46Q-r10-25:-N6QY2cz5Plg-ngRD6vW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6592b671d31ab9ca4af2e622ef81bee76c37721/pkg/sql/zoneconfig/zone_config_reader.go#L25)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

yeah, just remove this file in the first commit. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi commented:
Opened: https://github.com/cockroachdb/cockroach/issues/84050 for tracking the roundtrip time
